File: Greedy/763-partition_labels.py
from collections import Counter



def partition_labels(s: str) -> list:
    res = []
    count = Counter(s)
    addr = {}
    for i,c in enumerate(s):
        if c in addr:
            addr[c].append(i)
        else:
            addr[c] = [i]
    
    lst = []
    added = set()
    for c in s:
        if c in added:
            continue
        loc = addr[c]
        item = (c, loc[0], loc[-1])
        lst.append(item)
        added.add(c)

    total = count[lst[0][0]]
    prev = lst[0][2]
    for i in range(1, len(lst)):
        item = lst[i]
        if item[1] < prev:
            total += count[item[0]]
            if item[2] > prev:
                prev = item[2]
        else:
            res.append(total)
            total = count[item[0]]
            prev = item[2]
    res.append(total)
    return res


def standard_solution(s: str) -> list:
    res = []
    last = {c:i for i,c in enumerate(s)}
    j = anchor = 0
    for i,c in enumerate(s):
        j = max(j, last[c])
        if i == j:
            res.append(i - anchor + 1)
            anchor = i + 1
    return res


s = "ababcbacadefegdehijhklij"
res = partition_labels(s)
print(res)


res = standard_solution(s)
print(res)

File: Greedy/763-partition_labels.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <string>
#include <map>

using namespace std;


vector<int> partitionLabels(string S) {
    map<char, int> last;
    vector<int> ans;
    int size = S.size();
    for (int i=0; i<size; i++) {
        last[S[i]] = i;
    }

    int j=0, anchor=0;
    for (int i=0; i<size; i++) {
        j = max(last.at(S[i]), j);
        if (i == j) {
            ans.push_back(i-anchor+1);
            anchor = i+1;
        }
    }
    return ans;
}



int main() {
    string s = "ababcbacadefegdehijhklij";
    vector<int> res = partitionLabels(s);
    for (int i=0; i<res.size(); i++) {
        cout << res[i] << " ";
    }
    return 0;
}
